### ISSUE TITLE
Sorting Issues

### ISSUE THREAD

**Issue by sven-7** [ZeroQI/Absolute-Series-Scanner#101]
I've been seeing a lot of problems pop-up for sorting.  I'm using the scanner that's part of the latest beta.  

See the below screenshots:

<img width="1115" alt="screen shot 2017-07-05 at 9 45 25 pm" src="https://user-images.githubusercontent.com/18091557/27892025-62ba8f88-61cb-11e7-8d3a-ba29fbf13dd9.png">
<img width="676" alt="screen shot 2017-07-05 at 9 43 02 pm" src="https://user-images.githubusercontent.com/18091557/27892024-62b68c58-61cb-11e7-8691-e1988eaab88a.png">

Armed LibrariansL The Book of Bantorra is showing as main title, but sort title is The Book of Bantorra.  Same with Letter Bee, which has a sort title of Tegami Batch: Letter Bee. I have my title set to TheTVDB, AniDB and Sort title as well.

**Comment by ZeroQI** [comment]
Serie Language Priority main, en, jap
T-EM 'title' AniDB,TheTVDB: Tatakau Shisho: The Book of Bantorra
T-EM 'title_sort' AniDB,TheTVDB: Tatakau Shisho: The Book of Bantorra
no pb with latest beta

**Comment by ZeroQI** [comment]
Please confirm if issue present with
https://github.com/ZeroQI/Hama.bundle/releases/download/Beta/Hama.bundle.2017-06-30.00h11.zip

**Comment by ZeroQI** [comment]
Please confirm if issue present with
https://github.com/ZeroQI/Hama.bundle/releases/download/Beta/Hama.bundle.2017-06-30.00h11.zip

**Comment by sven-7** [comment]
Issue persists.

**Comment by sven-7** [comment]
Issue persists.

**Comment by sven-7** [comment]
I've looked into this some more.  My settings:

<img width="565" alt="screen shot 2017-08-15 at 10 42 46 am" src="https://user-images.githubusercontent.com/18091557/29320845-8862dd94-81a6-11e7-9ee3-d3f6890091f6.png">

Example:  Mouryou no Hako

HAMA:  [com.plexapp.agents.hama.log.zip](https://github.com/ZeroQI/Absolute-Series-Scanner/files/1225399/com.plexapp.agents.hama.log.zip)

ASS log:  [Mouryou no Hako.log.zip](https://github.com/ZeroQI/Absolute-Series-Scanner/files/1225376/Mouryou.no.Hako.log.zip)

In TVDB, [it does not have an alternate title](http://thetvdb.com/?tab=series&id=83474&lid=7), yet in Plex the 'Sort Title' says 'Box of Goblins'. AniDB [does have the alternate](http://anidb.net/perl-bin/animedb.pl?show=anime&aid=6047).  AniDB XML:

<img width="689" alt="screen shot 2017-08-15 at 10 42 24 am" src="https://user-images.githubusercontent.com/18091557/29320884-aef4dc14-81a6-11e7-8f80-2350dd2190ac.png">


Shows I have also noticed this happening:  

Space Pirate Captain Harlock (being sorted as Captain Harlock)
Monogatari (being sorted as Bakemonogatari)
Letter Bee (being sorted as Tegami Bachi: Letter Bee)

**Comment by sven-7** [comment]
Additionally, how does the scanner handle shows with 'The' in the front.  I have some shows that are filed under 'The [title]' and some that are not.

**Comment by ZeroQI** [comment]
```
  if string.endswith(", The"):                                   string = "The " + ''.join( string.split(", The", 1) )                                        # ", The" is rellocated in front
  if string.endswith(", A"  ):                                   string = "A "   + ''.join( string.split(", A"  , 1) )          
```
The scanner re-puts the The in front.
Title is re-applied by the agent anyway so need to check agent behaviour
I believe the title source is responsible. Can i have an exemple of each ?

**Comment by sven-7** [comment]
Sure. 

"The Disastrous Like of Saiki K" sorts as "Disastrous Like of Saiki K".  

<img width="684" alt="screen shot 2017-08-18 at 8 31 42 am" src="https://user-images.githubusercontent.com/18091557/29459089-25eca806-83f0-11e7-94f8-21d780e160b2.png">

The Asterisk War sorts as "The Asterisk War" instead of Asterisk War.

<img width="557" alt="screen shot 2017-08-18 at 8 37 53 am" src="https://user-images.githubusercontent.com/18091557/29459203-98982ab0-83f0-11e7-9b73-0672fe4f6f45.png">

I only have a few shows that start with 'A' and they are all sorted under A.

<img width="635" alt="screen shot 2017-08-18 at 8 37 07 am" src="https://user-images.githubusercontent.com/18091557/29459216-a7f1cbd8-83f0-11e7-9834-52fe79813676.png">


Below is an image of a lot of shows that have "The" in the title and then are sorted by "The" as well. 


<img width="1773" alt="screen shot 2017-08-18 at 8 32 48 am" src="https://user-images.githubusercontent.com/18091557/29459095-27fc9d40-83f0-11e7-850e-2ac85c0cab75.png">

**Comment by sven-7** [comment]
Note regarding #102 , when doing Monogatari via the AniDB2 it sorts as Bakemonogatari, but displays Monogatari.

<img width="675" alt="screen shot 2017-08-18 at 8 39 51 am" src="https://user-images.githubusercontent.com/18091557/29459262-ddab78be-83f0-11e7-8a39-0717e9f6923f.png">

**Comment by ZeroQI** [comment]
I believe sort title should not be by the "A " or "The " and i need to see why the different sort title. can you give me the agent and sort title priority ?

**Comment by sven-7** [comment]
<img width="524" alt="screen shot 2017-08-18 at 1 31 46 pm" src="https://user-images.githubusercontent.com/18091557/29470252-a4dc88b0-8419-11e7-9bdb-a1b44c7203f3.png">

Here are the logs for some of them too:

[Archive.zip](https://github.com/ZeroQI/Absolute-Series-Scanner/files/1235163/Archive.zip)

**Comment by ZeroQI** [comment]
i asked the question wrong my bad...
What language priority for "Serie Language Priority"?
you can use main instead of x-jat and replace jap with ja for that and "Episode Language Priority"

**Comment by sven-7** [comment]
Series language priority is: 'en' only.  I've added others after it, but it does not make a difference for sorting. I've updated it to: "en, main, x-jat, jap" with no difference. 

I think making it an Agent option would be great.  It gives everyone more custom features for how they want to curate their libraries.

**Comment by ZeroQI** [comment]
"en, main, ja" since jap doesn't exist put ja does (just noticed recently) and main = x-jat...
Will load en anidb title only. need to addd option but missing eps kept me busy, but found huge far reaching bug

**Comment by sven-7** [comment]
Okay, I usually have it set to 'en' only.  Will wait for the next update. Thanks!

**Comment by ZeroQI** [comment]
Working on it. In short:

Settings:
- Serie Language Priority: en
- T-EM 'title': TheTVDB, AniDB
- T-EM 'title_sort': TheTVDB, AniDB

Series showing issues:
- title: Tatakau Shisho: The Book of Bantorra, title_sort: Tatakau Shisho: The Book of Bantorra
- title: Monogatari, title_sort: Bakemonogatari
- title: Space Pirate Captain Harlock, title_sort: Bakemonogatari
- title: Letter Bee, title_sort: Tegami Bachi: Letter Bee
- title: Kurozuka, title_sort: Black Tomb
- title: Armed Librarians: The Book of Bantorra, title_sort: The Book of Bantorra.
- title: Mouryou no Hako, title_sort: Box of Goblins
- title: The Asterisk Wars: The Academy City On The Water, title_sort: The Asterisk Wars

Plex.UpdateMeta() - Metadata Fields (items #), type, source provider, value, 
[1] title                          Type: str                   Source: TheTVDB      Value: 'The Asterisk War: The Academy City on the Water'
[1] original_title                 Type: str                   Source: AniDB        Value: 'The Asterisk War'
[2] title_sort                     Type: str                   Source: AniDB        Value: 'The Asterisk War'

tvdb has title, not title sort, while anidb has same for both but different from thetvdb, so that match the code...

**Comment by sven-7** [comment]
> original_title doesn't update so sort title can be used to search 2 separate soure and/or language titles at the cost of the listing order

Does that mean this issue will persist?  If TheTVDB is set to priority in `'T-EM 'title_sort': TheTVDB, AniDB'` could it mimic the 'title' field?

I think an Agent option to allow sorting with/without articles would be good. Personally, I will sort without them.

**Comment by ZeroQI** [comment]
that mean people get to chose:
- same sort title cleaned or not for sorting depending agent option
- another title language/source that allow to search both sort and normal title but with incorrect sorting if they care more about 2 titles. this possibility will meaningless if plex fix original_title

here the beautiful short function. Now need to use regex:
```
def SortTitle(title, language=""):
  dict_sort = { 'en': ["The ", "A ", "An "]
                'fr': ["Le ", "La ", "Les ", "L ", "Un ", "Une ", "Des "]
                'sp': ["El ", "La ", "Las ", "Lo ", "Los ", "Uno ", "Una "]
              }
  prefix = title.replace("'", " ").split(" ", 1)[0]
  return title.replace(prefix, "", 1) if language in dict_sort and prefix in dict_sort[language] else title 
```
have to add agent option and launch from the right place now, will post when done

**Comment by sven-7** [comment]
Great. Will test when it's ready.

**Comment by ZeroQI** [comment]
uploaded. worked for The Asterisk War: The Academy City on the Water
agent option seem to work too

**Comment by sven-7** [comment]
Worked for me as well. Thanks again!

**Comment by sven-7** [comment]
Closed #101
